src/Data/HashMap/Linear.hs

```diff
@@ -59,218 +58,185 @@ import qualified Prelude
+empty sz@(Size s) scopeFn =
+  alloc s (-1, (error "key err val", error "val err val")) robinArrayScopeFn
   where
-    unsafeAlter :: Keyed k =>
-      HashMap k v -> (Maybe v -> Maybe v) -> k -> HashMap k v
-    unsafeAlter hmap f k =
-      case lookup hmap k of
-        (hmap', Unrestricted maybeV) -> case f maybeV of
-          Nothing -> delete hmap' k
-          Just v -> insert hmap' k v
-
--- | If the key is present, this update the value,
--- otherwise insert a new key-value pair.
-insert :: Keyed k => HashMap k v #-> k -> v -> HashMap k v
-insert hmap k v = insertFromQuery v $ queryIndex (maybeResize hmap) k
-  where
-    insertFromQuery :: Keyed k =>
-      v -> (HashMap k v, RobinQuery k) #-> HashMap k v
-    insertFromQuery v (HashMap (!size, !count) arr, IndexToInsert (k, psl) ix) =
-      HashMap (size, count + 1) (write arr ix (k, v, psl))
-    insertFromQuery v (HashMap sizes arr, IndexToUpdate (k, psl) ix) =
-      HashMap sizes (write arr ix (k, v, psl))
-    insertFromQuery v (HashMap sizes arr, IndexToSwap (k, psl) ix) =
-      (Unsafe.toLinear swapFromRead) (read arr ix) sizes ix (k, v, psl)
-    -- XXX: This re-does the probe sequence of (k',v')
-    swapFromRead ::
-      Keyed k =>
-      (RobinArr k v, RobinVal k v) ->
-      (Int, Int) ->
-      Int ->
-      RobinVal k v ->
-      HashMap k v
-    swapFromRead (arr', (k', v', _)) sizes ix (k, v, psl) =
-      insert (HashMap sizes (write arr' ix (k, v, psl))) k' v'
-
--- | This deletes the key-value pair if it is present, and otherwise does
--- nothing.
+    robinArrayScopeFn :: RobinArr k v #-> Unrestricted b
+    robinArrayScopeFn arr = scopeFn $ HashMap sz 0 arr
+
+-- | If the key is present, this update the value.
+-- If not, if there's enough space, insert a new pair.
+-- If there isn't enough space, error.
+insert :: (Keyed k, HasCallStack) => HashMap k v #-> k -> v -> HashMap k v
+insert (HashMap sz@(Size s) ct@(Count c) arr) k v = case c < s of
+  False -> lseq arr (error "Inserting into a full hashmap")
+  True -> tryInsertAtIndex (HashMap sz ct arr) ((hash k) `mod` s) (k,0) v
+
+-- | Try to insert at a given index with a given PSL. So the
+-- probing starts from the given index (with the given PSL).
+tryInsertAtIndex :: Keyed k =>
+  HashMap k v #-> Int -> (k, PSL) -> v -> HashMap k v
+tryInsertAtIndex hmap ix kPSL v = probeFrom kPSL ix hmap & \case
+  (HashMap sz ct arr, IndexToUpdate (k',psl') ix') ->
+    HashMap sz ct (write arr ix' (psl', (k',v)))
+  (HashMap sz ct arr, IndexToInsert (k',psl') ix') ->
+    HashMap sz (ct + 1) (write arr ix' (psl', (k',v)))
+  (HashMap (Size s) ct arr, IndexToSwap (k',psl') ix') -> read arr ix' & \case
+    (arr'', Unrestricted (p_old, (k_old, v_old))) ->
+      tryInsertAtIndex
+        (HashMap (Size s) ct (write arr'' ix' (psl', (k',v))))
+        ((ix' + 1) `mod` s)
+        (k_old, p_old + 1)
+        v_old
+
+-- | This deletes the key-value pair if it is present.
+-- If it's not present, this does nothing.
```

Comment:
When we are done with the review. I'll ask you do do a pass on the Haddock comments to remove all these “this”, and, generally, make the doc strings more precise and explicit. This can be a separate PR.